File: box-and-box/ports/elixir/lib/box_and_box/value.ex
defmodule BoxAndBox.Value do
  @moduledoc """
  Invariant Arithmetic — faithful port of value.mjs (v0.2).

  A Value is a PRODUCT OF MONOIDS across families. `combine/2` merges componentwise;
  `chain/2` composes along PULSE phases (partial — refuses a backward step);
  promote/reconcile/deliberate are endomorphisms; consume is the boolean gate.
  Laws L1–L14.

  Values are plain maps. Missing optional fields default like the JS `??`.
  """

  @phases ~w(retrieve route act learn consolidate)

  def phases, do: @phases

  def phase_idx(p), do: Enum.find_index(@phases, &(&1 == p)) || -1

  # identity element of the whole product monoid
  def v0 do
    %{
      n: 0,
      kappa: false,
      beta: 1,
      sigma: [],
      pi: nil,
      iota: nil,
      psi: nil,
      authority: [],
      deny_default: true,
      audit: []
    }
  end

  @doc "Build a Value with sensible defaults, copying list fields."
  def v(p \\ %{}) do
    base = v0()

    merged = Map.merge(base, p)

    %{
      merged
      | sigma: Enum.to_list(Map.get(p, :sigma, [])),
        authority: Enum.to_list(Map.get(p, :authority, [])),
        audit: Enum.to_list(Map.get(p, :audit, []))
    }
  end

  defp uniq(list), do: Enum.uniq(list)

  defp first_non_null(a, b), do: if(a != nil, do: a, else: b)

  # ---- combine : Value × Value → Value ----
  def combine(a, b) do
    %{
      n: a.n + b.n,
      kappa: a.kappa or b.kappa,
      beta: min(a.beta, b.beta),
      sigma: uniq(a.sigma ++ b.sigma),
      pi: first_non_null(a.pi, b.pi),
      iota: first_non_null(a.iota, b.iota),
      psi: first_non_null(a.psi, b.psi),
      authority: a.authority ++ b.authority,
      deny_default: a.deny_default and b.deny_default,
      audit: a.audit ++ b.audit
    }
  end

  # ---- chain : Value × Value → Value (PARTIAL) ----
  # Returns %{error: ...} on a backward phase step.
  def chain(a, b) do
    if a.pi != nil and b.pi != nil and phase_idx(a.pi) > phase_idx(b.pi) do
      %{error: "π-violation: cannot chain '#{b.pi}' after '#{a.pi}'"}
    else
      r = combine(a, b)
      %{r | pi: first_non_null(b.pi, a.pi)}
    end
  end

  # ---- promote : Value × Evidence → Value ----
  def promote(v, evidence \\ %{}) do
    ev_beta = Map.get(evidence, :beta, 0)
    %{v | beta: max(v.beta, ev_beta)}
  end

  # ---- reconcile : Value × Set<Tag> → Value ----
  def reconcile(v, tags \\ []) do
    drop = MapSet.new(tags)
    %{v | sigma: Enum.reject(v.sigma, &MapSet.member?(drop, &1))}
  end

  # ---- deliberate : Value → Value ----
  def deliberate(v), do: %{v | kappa: false}

  # ---- consume : Value × Requirements → {ok, failures, value} ----
  def consume(v, req \\ %{}) do
    failures =
      []
      |> beta_min(v, req)
      |> sigma_empty(v, req)
      |> acyclic(v, req)
      |> phase(v, req)
      |> forward_from(v, req)
      |> deny(v, req)
      |> Enum.reverse()

    %{ok: failures == [], failures: failures, value: v}
  end

  defp beta_min(acc, v, req) do
    bm = Map.get(req, :beta_min)

    if bm != nil and v.beta < bm,
      do: [%{family: "β", why: "β=#{round3(v.beta)} < β_min=#{bm}"} | acc],
      else: acc
  end

  defp sigma_empty(acc, v, req) do
    if Map.get(req, :sigma_empty) && length(v.sigma) > 0,
      do: [%{family: "σ", why: "unresolved conflicts {#{Enum.join(v.sigma, ", ")}}"} | acc],
      else: acc
  end

  defp acyclic(acc, v, req) do
    if Map.get(req, :acyclic) && v.kappa,
      do: [%{family: "κ", why: "cyclic — self-reference detected"} | acc],
      else: acc
  end

  defp phase(acc, v, req) do
    rp = Map.get(req, :phase)

    if rp != nil and v.pi != rp,
      do: [%{family: "π", why: "phase #{v.pi} ≠ required #{rp}"} | acc],
      else: acc
  end

  defp forward_from(acc, v, req) do
    ff = Map.get(req, :forward_from)

    if ff != nil and v.pi != nil and phase_idx(v.pi) < phase_idx(ff),
      do: [%{family: "π", why: "phase #{v.pi} precedes #{ff}"} | acc],
      else: acc
  end

  defp deny(acc, v, req) do
    if Map.get(req, :deny_default) == "must_allow" and v.deny_default == true and
         Map.get(req, :authorized) != true,
       do: [%{family: "governance", why: "deny_default with empty authority_path"} | acc],
       else: acc
  end

  def round3(x) when is_number(x), do: Float.round(x * 1.0, 3)
end

File: box-and-box/ports/rust/src/value.rs
// value.rs — Invariant Arithmetic (faithful port of value.mjs).
// A Value is a PRODUCT OF MONOIDS across families. combine merges; chain composes
// along PULSE phases (partial — refuses a backward step); promote/reconcile/deliberate
// are endomorphisms; consume is the boolean gate. Laws L1–L14.

pub const PHASES: [&str; 5] = ["retrieve", "route", "act", "learn", "consolidate"];

pub fn phase_idx(p: &str) -> i64 {
    PHASES.iter().position(|&x| x == p).map(|i| i as i64).unwrap_or(-1)
}

#[derive(Clone, Debug, PartialEq)]
pub struct Value {
    pub n: f64,                 // ℝ under +
    pub kappa: bool,            // Bool under ∨ — cyclicity
    pub beta: f64,              // [0,1] under min — persistence / confidence
    pub sigma: Vec<String>,     // Set<Tag> under ∪ — derived conflicts
    pub pi: Option<String>,     // Phase|null, first-non-null
    pub iota: Option<String>,   // IdemKey, first-non-null
    pub psi: Option<String>,    // Cadence, first-non-null
    pub authority: Vec<String>, // List<Cap> under concat (free monoid)
    pub deny_default: bool,     // Bool under ∧
    pub audit: Vec<String>,     // List<Event> under concat (free monoid)
}

// uniq preserving first-seen order (mirrors JS [...new Set(arr)])
fn uniq(items: &[String]) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    for it in items {
        if !out.contains(it) {
            out.push(it.clone());
        }
    }
    out
}

fn first_non_null(a: &Option<String>, b: &Option<String>) -> Option<String> {
    match a {
        Some(_) => a.clone(),
        None => b.clone(),
    }
}

impl Value {
    // identity element of the whole product monoid (V0)
    pub fn v0() -> Value {
        Value {
            n: 0.0,
            kappa: false,
            beta: 1.0,
            sigma: Vec::new(),
            pi: None,
            iota: None,
            psi: None,
            authority: Vec::new(),
            deny_default: true,
            audit: Vec::new(),
        }
    }
}

// ---- combine : Value × Value → Value ----
// componentwise monoid op. Associative with identity V0 ⇒ a monoid.
pub fn combine(a: &Value, b: &Value) -> Value {
    let mut sig = a.sigma.clone();
    sig.extend(b.sigma.iter().cloned());
    let mut auth = a.authority.clone();
    auth.extend(b.authority.iter().cloned());
    let mut aud = a.audit.clone();
    aud.extend(b.audit.iter().cloned());
    Value {
        n: a.n + b.n,
        kappa: a.kappa || b.kappa,
        beta: a.beta.min(b.beta),
        sigma: uniq(&sig),
        pi: first_non_null(&a.pi, &b.pi),
        iota: first_non_null(&a.iota, &b.iota),
        psi: first_non_null(&a.psi, &b.psi),
        authority: auth,
        deny_default: a.deny_default && b.deny_default,
        audit: aud,
    }
}

// chain result is either a Value or a π-violation error
pub enum ChainResult {
    Ok(Value),
    Error(String),
}

// ---- chain : Value × Value → Value (PARTIAL) ----
// Defined only when phase(a) ≤ phase(b); a backward step is REFUSED.
pub fn chain(a: &Value, b: &Value) -> ChainResult {
    if let (Some(pa), Some(pb)) = (&a.pi, &b.pi) {
        if phase_idx(pa) > phase_idx(pb) {
            return ChainResult::Error(format!(
                "π-violation: cannot chain '{}' after '{}'",
                pb, pa
            ));
        }
    }
    let mut r = combine(a, b);
    r.pi = first_non_null(&b.pi, &a.pi); // exit phase
    ChainResult::Ok(r)
}

// ---- promote : Value × Evidence → Value ----
// β-monotone endomorphism: promote(v).β ≥ v.β, always.
pub fn promote(v: &Value, evidence_beta: f64) -> Value {
    let mut r = v.clone();
    r.beta = v.beta.max(evidence_beta);
    r
}

// ---- reconcile : Value × Set<Tag> → Value ----
// σ-antitone, idempotent endomorphism: removes resolved conflict tags.
pub fn reconcile(v: &Value, tags: &[String]) -> Value {
    let mut r = v.clone();
    r.sigma = v.sigma.iter().filter(|t| !tags.contains(t)).cloned().collect();
    r
}

// ---- deliberate : Value → Value ----
// κ-antitone, idempotent endomorphism: forces κ = false.
pub fn deliberate(v: &Value) -> Value {
    let mut r = v.clone();
    r.kappa = false;
    r
}

#[derive(Clone, Debug)]
pub struct Failure {
    pub family: String,
    pub why: String,
}

#[derive(Clone, Debug, Default)]
pub struct Requirements {
    pub beta_min: Option<f64>,
    pub sigma_empty: bool,
    pub acyclic: bool,
    pub phase: Option<String>,
    pub forward_from: Option<String>,
    pub deny_default_must_allow: bool,
    pub authorized: bool,
}

pub struct ConsumeResult {
    pub ok: bool,
    pub failures: Vec<Failure>,
}

// ---- consume : Value × Requirements → {ok, failures} ----
pub fn consume(v: &Value, req: &Requirements) -> ConsumeResult {
    let mut failures: Vec<Failure> = Vec::new();
    if let Some(bm) = req.beta_min {
        if v.beta < bm {
            failures.push(Failure {
                family: "β".into(),
                why: format!("β={} < β_min={}", round(v.beta), bm),
            });
        }
    }
    if req.sigma_empty && !v.sigma.is_empty() {
        failures.push(Failure {
            family: "σ".into(),
            why: format!("unresolved conflicts {{{}}}", v.sigma.join(", ")),
        });
    }
    if req.acyclic && v.kappa {
        failures.push(Failure {
            family: "κ".into(),
            why: "cyclic — self-reference detected".into(),
        });
    }
    if let Some(ph) = &req.phase {
        if v.pi.as_deref() != Some(ph.as_str()) {
            failures.push(Failure {
                family: "π".into(),
                why: format!("phase {:?} ≠ required {}", v.pi, ph),
            });
        }
    }
    if let Some(ff) = &req.forward_from {
        if let Some(p) = &v.pi {
            if phase_idx(p) < phase_idx(ff) {
                failures.push(Failure {
                    family: "π".into(),
                    why: format!("phase {} precedes {}", p, ff),
                });
            }
        }
    }
    if req.deny_default_must_allow && v.deny_default && !req.authorized {
        failures.push(Failure {
            family: "governance".into(),
            why: "deny_default with empty authority_path".into(),
        });
    }
    ConsumeResult {
        ok: failures.is_empty(),
        failures,
    }
}

pub fn round(x: f64) -> f64 {
    (x * 1000.0).round() / 1000.0
}
